internal/batches/executor/task_status.go

```diff
@@ -8,18 +8,62 @@ import (
 	"github.com/sourcegraph/src-cli/internal/batches"
 )
 
+// TaskStatusCollection is a collection of TaskStatuses that provides
+// concurrency-safe access to the statuses by locking access and providing
+// copies of the TaskStatuses.
+type TaskStatusCollection struct {
+	statuses   map[*Task]*TaskStatus
+	statusesMu sync.RWMutex
+}
+
+func NewTaskStatusCollection(tasks []*Task) *TaskStatusCollection {
+	tsc := &TaskStatusCollection{
+		statuses: make(map[*Task]*TaskStatus),
+	}
+
+	for _, t := range tasks {
+		tsc.statuses[t] = &TaskStatus{
+			RepoName: t.Repository.Name,
+			Path:     t.Path,
+		}
+	}
+
+	return tsc
+}
+
+// Update updates the TaskStatus for the given Task by calling the provided
+// update callback with the TaskStatus.
+func (tsc *TaskStatusCollection) Update(task *Task, update func(status *TaskStatus)) {
+	tsc.statusesMu.Lock()
+	defer tsc.statusesMu.Unlock()
+
+	status, ok := tsc.statuses[task]
+	if ok {
+		update(status)
+	}
+}
+
+// CopyStatuses creates a copy of all TaskStatuses and calls the provided
+// callback with list.
+func (tsc *TaskStatusCollection) CopyStatuses(callback func([]*TaskStatus)) {
+	tsc.statusesMu.RLock()
+	defer tsc.statusesMu.RUnlock()
+
+	var s []*TaskStatus
+	for _, status := range tsc.statuses {
+		s = append(s, status)
+	}
```

Comment:
Do we need to shallow copy the statuses, since the read lock is still going to be held until after `callback(s)` returns anyway?

Reply:
This is just copy and pasted from the old method, so I didn't chnage it. I'm not 100% sure we still need it. Will investigate in a separate branch (because I'm scared of this part of the codebase)

Reply:
Ah, the reason we do the copy of the pointers is because `tsc.statuses` is a `map[*Task]*TaskStatus`, not a slice.